onnxruntime/core/common/cpuid_info.cc

```diff
@@ -266,8 +267,29 @@ void CPUIDInfo::ArmWindowsInit() {
   }
 }
 
-#endif /* (arm or arm64) and windows */
-#endif /* arm or arm64*/
+#elif defined(__APPLE__)  // ^ defined(_WIN32)
+
+void CPUIDInfo::ArmAppleInit() {
+#if defined(CPUINFO_SUPPORTED)
+  if (pytorch_cpuinfo_init_) {
+    is_hybrid_ = cpuinfo_get_uarchs_count() > 1;
+    has_arm_neon_dot_ = cpuinfo_has_arm_neon_dot();
+    has_fp16_ = cpuinfo_has_arm_neon_fp16_arith();
+    has_arm_neon_i8mm_ = cpuinfo_has_arm_i8mm();
+    has_arm_sve_i8mm_ = cpuinfo_has_arm_sve() && cpuinfo_has_arm_i8mm();
+    has_arm_neon_bf16_ = cpuinfo_has_arm_neon_bf16();
+
+    // TODO figure out how to set core_uarchs_ and is_armv8_narrow_ld_
```

Comment:
FWIW it looks like cpuinfo has some newer apple uarch values than common/cpuid_uarch.h. Should we update that list. 

https://github.com/pytorch/cpuinfo/blob/3c8b1533ac03dd6531ab6e7b9245d488f13a82a5/include/cpuinfo.h#L566-L573

but still lacks info for the latest apple chips.

Would we make any different decisions if the uarch values were set. e.g. would that give better info on big/little cores to use?

Would the value of is_armv8_narrow_ld_ differ across the various apple chips? Seemed like a small subset of Cortex chips had that set so naively wondering if a simple default value would be sufficient.

Reply:
good questions, I'm not too clear on the answers yet.
 
@chenfucn do you know?

Reply:
will get an initial implementation in for 1.18.1. we can refine this once we know how to in a future PR.